rpc/gnmi/gnmi-commit-confirmed.md

```diff
@@ -0,0 +1,106 @@
+client workflows and server implementation.
+
+The server can be viewed as a singleton resource, at any given point in time there
+can be only one commit active. This commit can be either confirmed or canceled
+before a new commit can begin. Client is expected to provide full configuration
+during the commit request, the commit cannot be amended once issued.
+
+# 3. Definition
+
+A `Commit` message is embedded the Extension message of the SetRequest proto.
+
+## 3.1 Proto
+
+```
+// Commit confirmed extension allows automated revert of the configuration after
+// certain duration if an explicit confirmation is not issued. It allows explicit
+// cancellation of the commit during the rollback window. There cannot be more
+// than one commit active at a given time.
+// The document about gNMI commit confirmed can be found at
+// https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-commit-confirmed.md
+message Commit {
+  // ID is provided by the client during the commit request. During confirm and cancel
+  // actions the provided ID should match the ID provided during commit.
+  // If ID is not passed in any actions server shall return error.
+  // Required.
+  string id = 1;
+  oneof action {
+    // commit action creates a new commit. If a commit is on-going, server returns error.
+    CommitRequest commit = 2;
+    // confirm action will confirm an on-going commit, the ID provided during confirm
+    // should match the on-going commit ID.
+    CommitConfirm confirm = 3;
+    // cancel action will cancel an on-going commit, the ID provided during cancel
+    // should match the on-going commit ID.
+    CommitCancel cancel = 4;
+  }
+}
+
+// CommitRequest is used to create a new confirmed commit. It hold additional
+// parameter requried for commit action.
+message CommitRequest {
+  // Maximum duration to wait for a confirmaton before reverting the commit.
+  google.protobuf.Duration rollback_duration = 1;
+}
+
+// CommitConfirm is used to confirm an on-going commit. It hold additional
+// parameter requried for confirm action.
+message CommitConfirm {}
+
+// CommitCancel is used to cancel an on-going commit. It hold additional
+// parameter requried for cancel action.
+message CommitCancel {}
+```
+
+## 3.2 SetRequest handling                                                        
+
+### 3.2.1 Commit
+A commit can be initiated by providing `CommitRequest` as action in the extension. A `id` must to be
+provided by the client. The server shall associate the commit with the provided `id`.
+During confirm or cancel action the provided `id` must match the `id` of the on-going commit.
+`rollback_duration` can be used to override the default rollback duration which is 10min.
+If a confirmation call is not received before the rollback duration then the configuration is reverted.
+
+If a commit is issued whilst an existing rollback counter is running then the server returns with
+FAILED_PRECONDITION error.
+
```

Comment:
@dplore @kidiyoor upon further review of the spec I wanted to call out one important point that I think is missing in this proposal - **Resetting the rollback timer**.

We would like to consider adding this option to the spec of the extension. 

For network wide commits we want to set a short timeout for fast recovery but extend it while waiting for slow nodes on the network.

In our original proposal we had the following definition of the resetting logic:

> To reset the timeout for a commit for which the timeout timer has not yet expired, the client MUST include the `CommitConfirmed` extension in the `SetRequest` message without any content in `update`/`delete`/`replace` fields. The `CommitConfirmed` message MUST include the `COMMIT_CONFIRMED_ACTION_START` action, `timeout` value set to the new desired timeout, and `commit_id` field set to the commit identifier returned by the network device in the respecting `SetResponse` message.

> The network device should reset the commit rollback timeout to the new value and return the `SetResponse` message without `CommitConfirmed` extension.

Reply:
This looks like an new action that can be added on top of the current proposal ? Could this be added in a follow-up PR ?

Reply:
While we can definitely do that in a separate PR I wanted to check why do you think this has to be done separately.

We can discuss if that warrants another message under `oneof`, or if it can be the same `CommitRequest` provided with the same ID and a new rollback timeout value.

Reply:
I was hoping we could do this in this PR, to keep the momentum.

The resetting timeout is in-line with prior art in netconf https://datatracker.ietf.org/doc/html/rfc6241#section-8.4

> If a follow-up confirmed <commit> operation is issued before the timer expires, the timer is reset to the new value (600 seconds by default).

We can argue if that is better to be modelled as another distinctive action, or be part of the `CommitRequest` message.

I can create a PR targeting this PR if required with the proposed additions.

Reply:
If additional changes can be made in subsequent proposal/PR we should consider that to unblock folks and avoid unnecessary delays for already approved changes - https://github.com/openconfig/gnmi/pull/155#pullrequestreview-1731027525. 

I suspect there will be some discussion and decision to be made in terms of the syntax of the `ResetTimer` cmd. Probably best to keep the discussion in separate PR as `ResetTimer` is not core functionality of `Confirmed Commit`, its an optimization.

Reply:
We can consider giving it a round of discussion at the next OC call, given how frictionless and low-engaging the past few calls went I do not expect this to bear any substantial discussions.

The reason I wanted to do this here, is that I think this rather belongs to the commit confirmed functionality, and therefore must be defined in the same spec file

Reply:
The reset rollback timer is being worked at https://github.com/openconfig/gnmi/pull/161